File: src/main/java/site/liangbai/realhomehunt/common/database/converter/LocationConverter.java
/*
 * RealHomeHunt
 * Copyright (C) 2021  Liangbai
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package site.liangbai.realhomehunt.common.database.converter;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;

import javax.persistence.AttributeConverter;
import javax.persistence.Converter;

@Converter
public class LocationConverter implements AttributeConverter<Location, String> {

    @Override
    public String convertToDatabaseColumn(Location attribute) {
        JsonObject jsonObject = new JsonObject();

        if (attribute.getWorld() != null) {
            jsonObject.addProperty("world", attribute.getWorld().getName());
        }

        jsonObject.addProperty("x", attribute.getX());
        jsonObject.addProperty("y", attribute.getY());
        jsonObject.addProperty("z", attribute.getZ());

        jsonObject.addProperty("yaw", attribute.getYaw());
        jsonObject.addProperty("pitch", attribute.getPitch());

        return jsonObject.toString();
    }

    @Override
    public Location convertToEntityAttribute(String dbData) {
        JsonObject jsonObject = new JsonParser().parse(dbData).getAsJsonObject();

        World world = null;
        if (jsonObject.has("world")) {
            String worldName = jsonObject.get("world").getAsString();
            world = Bukkit.getWorld(worldName);
            if (world == null) {
                throw new IllegalArgumentException("unknown world: " + worldName);
            }
        }

        double x = jsonObject.get("x").getAsDouble();
        double y = jsonObject.get("y").getAsDouble();
        double z = jsonObject.get("z").getAsDouble();

        float yaw = jsonObject.get("yaw").getAsFloat();
        float pitch = jsonObject.get("pitch").getAsFloat();

        return new Location(world, x, y, z, yaw, pitch);
    }
}

File: src/main/java/site/liangbai/realhomehunt/common/database/converter/LocationConverter.kt
/*
 * RealHomeHunt
 * Copyright (C) 2021  Liangbai
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package site.liangbai.realhomehunt.common.database.converter

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import javax.persistence.AttributeConverter
import javax.persistence.Converter

@Converter
class LocationConverter : AttributeConverter<Location, String> {
    companion object {
        fun Location.asJsonObject(): JsonObject {
            return JsonObject()
                .also {
                    if (world != null) {
                        it.addProperty("world", world!!.name)
                    }
                    it.addProperty("x", x)
                    it.addProperty("y", y)
                    it.addProperty("z", z)
                    it.addProperty("yaw", yaw)
                    it.addProperty("pitch", pitch)
                }
        }
    }

    override fun convertToDatabaseColumn(attribute: Location): String {
        return attribute.asJsonObject().toString()
    }

    override fun convertToEntityAttribute(dbData: String): Location {
        val jsonObject = JsonParser().parse(dbData).asJsonObject
        var world: World? = null
        if (jsonObject.has("world")) {
            val worldName = jsonObject["world"].asString
            world = Bukkit.getWorld(worldName)
            requireNotNull(world) { "unknown world: $worldName" }
        }
        val x = jsonObject["x"].asDouble
        val y = jsonObject["y"].asDouble
        val z = jsonObject["z"].asDouble
        val yaw = jsonObject["yaw"].asFloat
        val pitch = jsonObject["pitch"].asFloat
        return Location(world, x, y, z, yaw, pitch)
    }
}